### <span style="color:pink; font-weight:bold"> CURSO - Python Pandas: tratando e analisando dados</span>

O Pandas é uma biblioteca do Python que oferece estruturas de dados e ferramentas projetadas para tornar mais fácil o processo de manipulação e análise de informações.  
Desenvolveremos um projeto em que faremos realmente o papel de um cientista de dados, participaremos do processo inicial de conhecimento, tratamento e limpeza de uma base de dados e utilizando o Pandas para resolver os problemas que se apresentarem. Lidaremos com uma seguradora fictícia que pretende elaborar produtos para o mercado imobiliário e para isso ela possui uma base de dados para ser trabalhada e desenvolver análises e relatórios. 

#### <span style="color:pink; font-weight:bold"> Conhecendo Jupyter</span>

O primeiro passo para iniciarmos nosso curso de Pandas é configurar o ambiente. Conheceremos a distribuição do Python, Anaconda, própria para Data Science por conter todo o conjunto de ferramentas de que precisamos para resolver problemas.  
O Anaconda ainda fornece uma IDE padrão denominada Spider, não usaremos este recurso no projeto, mas sim o Jupyter, uma ferramenta mais interativa. 

##### <span style="color:pink; font-weight:bold"> Ambiente Virtual</span>

Porque criar ambientes virtuais? Imagine que há seis meses você escreveu um código complicado e longo usando várias bibliotecas e pacotes para criar uma base de dados. Agora, o seu chefe quer atualizar essa base de dados e, de repente, o seu código já não funciona. Isso acontece porque as bibliotecas foram atualizadas, o que pode levar a mudanças ou até mesmo a exclusão de algumas funcionalidades.  

No entanto, se você cria um ambiente virtual, a manutenção do código se torna mais fácil. Você pode especificar a versão da biblioteca que deseja usar, garantindo que o código funcione da mesma maneira que antes da atualização, mesmo que as bibliotecas tenham mudado.

#### <span style="color:pink; font-weight:bold"> Importando dados</span>

##### <span style="color:pink; font-weight:bold"> Relatório de análise 1</span>

In [ ]:
import pandas as pd

In [ ]:
pd.read_csv('/home/ba/Documentos/data-science/curso2_python_pandas/aluguel.csv', sep=';')

In [ ]:
dados = pd.read_csv('/home/ba/Documentos/data-science/curso2_python_pandas/aluguel.csv', sep=';')

In [ ]:
type(dados)

In [ ]:
dados.info()

In [ ]:
dados.head()

In [ ]:
dados.dtypes

In [ ]:
tipos_de_dados = pd.DataFrame(dados.dtypes, columns= ['Tipos de Dados']) #nomeando colunas

In [ ]:
tipos_de_dados.columns.name = 'Variáveis'

In [ ]:
tipos_de_dados

In [ ]:
dados.shape

In [ ]:
dados.shape[0]

In [ ]:
dados.shape[1]

In [ ]:
print('A base de dados apresenta {} registros (imóveis) e {} variáveis'.format(dados.shape[0], dados.shape[1]))

In [ ]:
# importando dados de arquivo json
import pandas as pd

json = open('/home/ba/Documentos/data-science/curso2_python_pandas/dados/aluguel.json')
print(json.read())

In [ ]:
df_json = pd.read_json('/home/ba/Documentos/data-science/curso2_python_pandas/dados/aluguel.json')
df_json

In [ ]:
# importando dados de arquivo txt
import pandas as pd

txt = open('/home/ba/Documentos/data-science/curso2_python_pandas/dados/aluguel.txt')
print(txt.read())

In [ ]:
df_txt = pd.read_table('/home/ba/Documentos/data-science/curso2_python_pandas/dados/aluguel.txt')
df_txt

In [ ]:
%pip install openpyxl

In [ ]:
df_xlsx = pd.read_excel('/home/ba/Documentos/data-science/curso2_python_pandas/dados/aluguel.xlsx')
df_xlsx

In [ ]:
%pip install lxml

In [ ]:
df_html = pd.read_html('dados/dados_html_1.html')
df_html[0]

#### <span style="color:pink; font-weight:bold"> Séries e Index</span>

##### <span style="color:pink; font-weight:bold"> Relatório de análise 2</span>

Identificar todos os tipos de imóveis que tem na nossa base

In [ ]:
import pandas as pd

dados = pd.read_csv('/home/ba/Documentos/data-science/curso2_python_pandas/aluguel.csv', sep=';')

In [ ]:
dados.head()

In [ ]:
dados['Tipo']

In [ ]:
tipo_de_imovel = dados['Tipo']

In [ ]:
type(tipo_de_imovel)

O tipo series trata-se de uma estrutura do Pandas, trata-se um array unidimensional. 

In [ ]:
tipo_de_imovel.drop_duplicates()  # assim conseguimos visualizar todos os imóveis que tem pois suas cópias foram excluidas

In [ ]:
tipo_de_imovel.drop_duplicates(inplace=True) #inplace já salva os valores na variavel

Configurando o parâmetro inplace como True, a modificação é realizada a partir da aplicação do próprio método, sem necessidade de criar um DataFrame para armazenar o resultado.

##### <span style="color:pink; font-weight:bold"> Organizando a visualização</span>

In [ ]:
tipo_de_imovel = pd.DataFrame(tipo_de_imovel)

In [ ]:
tipo_de_imovel.index

In [ ]:
tipo_de_imovel.shape[0]

In [ ]:
# enumerando bonitinho
range(tipo_de_imovel.shape[0])

In [ ]:
for i in range(tipo_de_imovel.shape[0]):
    print(i)

In [ ]:
tipo_de_imovel.index = range(tipo_de_imovel.shape[0])

In [ ]:
tipo_de_imovel.columns.name = 'Id'

In [ ]:
tipo_de_imovel

##### <span style="color:pink; font-weight:bold"> Series</span>

In [ ]:
data = [1,2,3,4,5]
s= pd.Series(data)

In [ ]:
s

In [ ]:
index = ['linha'+str(i) for i in range(5)]

In [ ]:
index

In [ ]:
s =pd.Series(data = data, index = index)

In [ ]:
data = {'Linha' + str(i) : i + 1 for i in range(5) }

In [ ]:
data

In [ ]:
s = pd.Series(data)
s

In [ ]:
data = [[1, 2, 3], 
        [4, 5, 6], 
        [7, 8, 9]]

In [ ]:
df1 = pd.DataFrame(data = data)

In [ ]:
df1


In [ ]:
index = ['Linha' + str(i) for i in range(3)]

In [ ]:
index

In [ ]:
df1 = pd.DataFrame(data = data, index = index)

In [ ]:
df1

In [ ]:
columns = ['Coluna' + str(i) for i in range(3)]
columns

In [ ]:
df1 = pd.DataFrame(data = data, index = index, columns=columns)

In [ ]:
df1

In [ ]:
data = {'Coluna0': {'Linha0': 1, 'Linha1': 4, 'Linha2': 7},
        'Coluna1': {'Linha0': 2, 'Linha1': 5, 'Linha2': 8},
        'Coluna2': {'Linha0': 3, 'Linha1': 6, 'Linha2': 9}}

In [ ]:
data

In [ ]:
df2 = pd.DataFrame(data)

In [ ]:
df2

é possível fazer a mesma coisa usando tuplas

In [ ]:
data = [(1, 2, 3), 
        (4, 5, 6), 
        (7, 8, 9)]

In [ ]:
df3 = pd.DataFrame(data=data, index=index, columns = columns)
df3

In [ ]:
df1[df1 > 0] = 'A'
df1

In [ ]:
df2[df2 > 0] = 'B'
df2

In [ ]:
df3[df3 > 0] = 'C'
df3

In [ ]:
df4 = pd.concat([df1,df2,df3])
df4

In [ ]:
df4 = pd.concat([df1,df2,df3], axis=1)
df4

#### <span style="color:pink; font-weight:bold"> Filtrando dados</span>

Com base no último relatório que entregamos, a empresa decidiu criar produtos apenas para imóveis residenciais. Dessa forma, precisamos criar uma nova base de dados que contenha somente imóveis dessa categoria.

##### <span style="color:pink; font-weight:bold"> Relatório de análise III</span>

###### <span style="color:pink; font-weight:bold"> Imóveis Residênciais</span>

In [75]:
import pandas as pd

dados = pd.read_csv('/home/ba/Documentos/data-science/curso2_python_pandas/aluguel.csv', sep=';')
dados.head(10)

Tipo           Bairro  Quartos  Vagas  Suites  Area  \
0                Quitinete       Copacabana        1      0       0    40   
1                     Casa  Jardim Botânico        2      0       1   100   
2  Conjunto Comercial/Sala  Barra da Tijuca        0      4       0   150   
3              Apartamento           Centro        1      0       0    15   
4              Apartamento     Higienópolis        1      0       0    48   
5              Apartamento     Vista Alegre        3      1       0    70   
6              Apartamento         Cachambi        2      0       0    50   
7       Casa de Condomínio  Barra da Tijuca        5      4       5   750   
8       Casa de Condomínio            Ramos        2      2       0    65   
9  Conjunto Comercial/Sala           Centro        0      3       0   695   

     Valor  Condominio    IPTU  
0   1700.0       500.0    60.0  
1   7000.0         NaN     NaN  
2   5200.0      4020.0  1111.0  
3    800.0       390.0    20.0  
4    800.0       230.0     NaN  
5   1200.0         NaN     NaN  
6   1300.0       301.0    17.0  
7  22000.0         NaN     NaN  
8   1000.0         NaN     NaN  
9  35000.0     19193.0  3030.0

Usaremos o drop.duplicate para remover todos os tipos iguais para nos auxiliar na seleção dos imóveis residenciais. Primeiro criaremos uma lista, agregamos nossos dados e então utilizaremos o método drop.duplicate.

In [76]:
list(dados['Tipo'].drop_duplicates())

['Quitinete',
 'Casa',
 'Conjunto Comercial/Sala',
 'Apartamento',
 'Casa de Condomínio',
 'Prédio Inteiro',
 'Flat',
 'Loja/Salão',
 'Galpão/Depósito/Armazém',
 'Casa Comercial',
 'Casa de Vila',
 'Terreno Padrão',
 'Box/Garagem',
 'Loft',
 'Loja Shopping/ Ct Comercial',
 'Chácara',
 'Loteamento/Condomínio',
 'Sítio',
 'Pousada/Chalé',
 'Studio',
 'Hotel',
 'Indústria']

In [77]:
residencial = ['Quitinete',
 'Casa',
 'Apartamento',
 'Casa de Condomínio',
 'Casa de Vila']

Usando o método isin() verificamos se os elementos dentro do nosso DataFrame da nossa lista existem de fato, com True e False.

In [78]:
dados['Tipo'].isin(residencial).head(10)

0     True
1     True
2    False
3     True
4     True
5     True
6     True
7     True
8     True
9    False
Name: Tipo, dtype: bool

Atribuiremos a nossa lista de imóveis residenciais a variável selecao.

In [79]:
selecao = dados['Tipo'].isin(residencial)

In [80]:
dados_residencial = dados[selecao]

Como resultado, temos apenas imóveis do tipo residencial:

In [81]:
dados_residencial

Tipo                Bairro  Quartos  Vagas  Suites  Area  \
0        Quitinete            Copacabana        1      0       0    40   
1             Casa       Jardim Botânico        2      0       1   100   
3      Apartamento                Centro        1      0       0    15   
4      Apartamento          Higienópolis        1      0       0    48   
5      Apartamento          Vista Alegre        3      1       0    70   
...            ...                   ...      ...    ...     ...   ...   
32953  Apartamento                 Méier        2      0       0    70   
32955    Quitinete                Centro        0      0       0    27   
32956  Apartamento           Jacarepaguá        3      1       2    78   
32957  Apartamento  São Francisco Xavier        2      1       0    48   
32958  Apartamento                Leblon        2      0       0    70   

        Valor  Condominio  IPTU  
0      1700.0       500.0  60.0  
1      7000.0         NaN   NaN  
3       800.0       390.0  20.0  
4       800.0       230.0   NaN  
5      1200.0         NaN   NaN  
...       ...         ...   ...  
32953   900.0       490.0  48.0  
32955   800.0       350.0  25.0  
32956  1800.0       800.0  40.0  
32957  1400.0       509.0  37.0  
32958  3000.0       760.0   NaN  

[22580 rows x 9 columns]

Agora verificando a veracidade dos nossos dados, vamos coletar a variável tipo e usaremos o méto drop_duplicates

In [82]:
list(dados_residencial['Tipo'].drop_duplicates())

['Quitinete', 'Casa', 'Apartamento', 'Casa de Condomínio', 'Casa de Vila']

CHecando o tamanho do DataFrame:

In [83]:
dados_residencial.shape[0]

22580

Agora vamos realizar a reconstrução do índice de dados_residencial. Podemos verificar que a numeração está incorreta pois está faltando o número 2 e 9 por exemplo, pois esses elementos foram excluidos.  
Usaremos o termo index e o método range que receberá os dados_residencial.shape[0]

In [84]:
dados_residencial.index = range(dados_residencial.shape[0])

In [85]:
dados_residencial

Tipo                Bairro  Quartos  Vagas  Suites  Area  \
0        Quitinete            Copacabana        1      0       0    40   
1             Casa       Jardim Botânico        2      0       1   100   
2      Apartamento                Centro        1      0       0    15   
3      Apartamento          Higienópolis        1      0       0    48   
4      Apartamento          Vista Alegre        3      1       0    70   
...            ...                   ...      ...    ...     ...   ...   
22575  Apartamento                 Méier        2      0       0    70   
22576    Quitinete                Centro        0      0       0    27   
22577  Apartamento           Jacarepaguá        3      1       2    78   
22578  Apartamento  São Francisco Xavier        2      1       0    48   
22579  Apartamento                Leblon        2      0       0    70   

        Valor  Condominio  IPTU  
0      1700.0       500.0  60.0  
1      7000.0         NaN   NaN  
2       800.0       390.0  20.0  
3       800.0       230.0   NaN  
4      1200.0         NaN   NaN  
...       ...         ...   ...  
22575   900.0       490.0  48.0  
22576   800.0       350.0  25.0  
22577  1800.0       800.0  40.0  
22578  1400.0       509.0  37.0  
22579  3000.0       760.0   NaN  

[22580 rows x 9 columns]

In [89]:
#exercicio
import pandas as pd

numeros = [i for i in range(11)]
letras = [chr(i + 65) for i in range(11)]
nome_coluna = ['N']

df = pd.DataFrame(data = numeros, index = letras, columns = nome_coluna)
df

N
A   0
B   1
C   2
D   3
E   4
F   5
G   6
H   7
I   8
J   9
K  10

In [90]:
#exercicio
selecao = df['N'].isin([i for i in range(11) if i % 2 == 0])
df = df[selecao]
df

N
A   0
C   2
E   4
G   6
I   8
K  10

Agora aprenderemos a exportar um dataframe e salvalo em formato csv. Temos o método read_csv() para importação, para exportação de dados utilizaremos to_csv()

In [92]:
dados_residencial.to_csv('dados/aluguel_residencial.csv', sep = ';')

dados_residencial2 = pd.read_csv('dados/aluguel_residencial.csv', sep = ';')
dados_residencial2 #a gente observa que ele exportou o indice também 

Unnamed: 0         Tipo                Bairro  Quartos  Vagas  Suites  \
0               0    Quitinete            Copacabana        1      0       0   
1               1         Casa       Jardim Botânico        2      0       1   
2               2  Apartamento                Centro        1      0       0   
3               3  Apartamento          Higienópolis        1      0       0   
4               4  Apartamento          Vista Alegre        3      1       0   
...           ...          ...                   ...      ...    ...     ...   
22575       22575  Apartamento                 Méier        2      0       0   
22576       22576    Quitinete                Centro        0      0       0   
22577       22577  Apartamento           Jacarepaguá        3      1       2   
22578       22578  Apartamento  São Francisco Xavier        2      1       0   
22579       22579  Apartamento                Leblon        2      0       0   

       Area   Valor  Condominio  IPTU  
0        40  1700.0       500.0  60.0  
1       100  7000.0         NaN   NaN  
2        15   800.0       390.0  20.0  
3        48   800.0       230.0   NaN  
4        70  1200.0         NaN   NaN  
...     ...     ...         ...   ...  
22575    70   900.0       490.0  48.0  
22576    27   800.0       350.0  25.0  
22577    78  1800.0       800.0  40.0  
22578    48  1400.0       509.0  37.0  
22579    70  3000.0       760.0   NaN  

[22580 rows x 10 columns]

In [93]:
dados_residencial.to_csv('dados/aluguel_residencial.csv', sep = ';', index=False)

In [95]:
dados_residencial2 = pd.read_csv('dados/aluguel_residencial.csv', sep = ';')
dados_residencial2 # agora foi exportado sem o índice

Tipo                Bairro  Quartos  Vagas  Suites  Area  \
0        Quitinete            Copacabana        1      0       0    40   
1             Casa       Jardim Botânico        2      0       1   100   
2      Apartamento                Centro        1      0       0    15   
3      Apartamento          Higienópolis        1      0       0    48   
4      Apartamento          Vista Alegre        3      1       0    70   
...            ...                   ...      ...    ...     ...   ...   
22575  Apartamento                 Méier        2      0       0    70   
22576    Quitinete                Centro        0      0       0    27   
22577  Apartamento           Jacarepaguá        3      1       2    78   
22578  Apartamento  São Francisco Xavier        2      1       0    48   
22579  Apartamento                Leblon        2      0       0    70   

        Valor  Condominio  IPTU  
0      1700.0       500.0  60.0  
1      7000.0         NaN   NaN  
2       800.0       390.0  20.0  
3       800.0       230.0   NaN  
4      1200.0         NaN   NaN  
...       ...         ...   ...  
22575   900.0       490.0  48.0  
22576   800.0       350.0  25.0  
22577  1800.0       800.0  40.0  
22578  1400.0       509.0  37.0  
22579  3000.0       760.0   NaN  

[22580 rows x 9 columns]

##### <span style="color:pink; font-weight:bold"> Relatório de análise IV</span>

Com base no novo banco de dados atualizado, a seguradora deseja que realizemos algumas seleções e retire a frequência de cada uma delas. Recebemos o seguinte e-mail da empresa:

    Selecione somente os imóveis classificados com tipo 'Apartamento'.

    Selecione os imóveis classificados com tipos 'Casa', 'Casa de Condomínio' e 'Casa de Vila'.

    Selecione os imóveis com área entre 60 e 100 metros quadrados, incluindo os limites.

    Selecione os imóveis que tenham pelo menos 4 quartos e aluguel menor que R$ 2.000,00.


In [96]:
dados = pd.read_csv('dados/aluguel_residencial.csv', sep = ';')

In [97]:
dados.head(10)

Tipo           Bairro  Quartos  Vagas  Suites  Area    Valor  \
0           Quitinete       Copacabana        1      0       0    40   1700.0   
1                Casa  Jardim Botânico        2      0       1   100   7000.0   
2         Apartamento           Centro        1      0       0    15    800.0   
3         Apartamento     Higienópolis        1      0       0    48    800.0   
4         Apartamento     Vista Alegre        3      1       0    70   1200.0   
5         Apartamento         Cachambi        2      0       0    50   1300.0   
6  Casa de Condomínio  Barra da Tijuca        5      4       5   750  22000.0   
7  Casa de Condomínio            Ramos        2      2       0    65   1000.0   
8         Apartamento           Centro        1      0       0    36   1200.0   
9         Apartamento           Grajaú        2      1       0    70   1500.0   

   Condominio  IPTU  
0       500.0  60.0  
1         NaN   NaN  
2       390.0  20.0  
3       230.0   NaN  
4         NaN   NaN  
5       301.0  17.0  
6         NaN   NaN  
7         NaN   NaN  
8         NaN   NaN  
9       642.0  74.0

Criaremos uma variável selecao e atribuiremos a ela a Series booleana, como fizemos nas aulas anteriores.

In [99]:
#Selecione somente os imóveis classificados com tipo 'Apartamento'.
selecao = dados['Tipo'] == 'Apartamento'
n1 = dados[selecao].shape[0]
n1 #saída da quantidade de apartamentos no banco de dados


19532

In [101]:
#Selecione os imóveis classificados com tipos 'Casa', 'Casa de Condomínio' e 'Casa de Vila'.

selecao = (dados['Tipo'] == 'Casa') | (dados['Tipo'] == 'Casa de Condomínio') | (dados['Tipo'] == 'Casa de Vila')
n2 = dados[selecao].shape[0]
n2

2212

Nossa próxima tarefa é selecionar imóveis que tenham a área entre 60 e 100 metros quadrados, isto é, estamos criando condições e precisaremos usar um novo operador, o &. Em termos matemáticos, o tópico poderia ser escrito da seguinte maneira: 60 <= Area <= 100, isto é, buscamos uma área que está entre 60 e 100 metros quadrados

Desta vez, trabalharemos com a variável Area , e como teremos dois temos neste código, então devemos usar () para separá-los. Um dos termos deve ser maior ou igual(>=) a 60, enquanto o outro deve ser mais ou igual a 100.

Por fim, verificaremos a quantidade de elementos no DataFrame. 

In [103]:
#Selecione os imóveis com área entre 60 e 100 metros quadrados, incluindo os limites.
#60 <= Area <=100
selecao = (dados['Area'] >= 60) & (dados['Area'] <= 100)
n3 = dados[selecao].shape[0]
n3

8719

A próxima tarefa requisitada pela seguradora é selecionarmos imóveis que contenham pelo menos quatro quartos e um aluguel menor que R$ 2.000,00. Trabalharemos com duas variáveis diferentes neste caso: Quartos e Valor, em seguida estipularemos uma quantidade específica para cada variável, de forma que sejam selecionados apenas os elementos que preenchem os requisitos. 

In [106]:
#Selecione os imóveis que tenham pelo menos 4 quartos e aluguel menor que R$ 2.000,00.

selecao = (dados['Quartos'] >= 4) & (dados['Valor'] <2000)
n4 = dados[selecao].shape[0]
n4

41

In [107]:
print("Nº de imóveis classificados com tipo 'Apartamento' -> {}".format(n1))
print("Nº de imóveis classificados com tipos 'Casa', 'Casa de Condomínio' e 'Casa de Vila'-> {}".format(n2))
print("Nº de imóveis com área entre 60 e 100 metros quadrados, incluindo os limites -> {}".format(n3))
print("Nº de imóveis que tenham pelo menos 4 quartos e aluguel menor que R$ 2.000,00 -> {}".format(n4))


Nº de imóveis classificados com tipo 'Apartamento' -> 19532
Nº de imóveis classificados com tipos 'Casa', 'Casa de Condomínio' e 'Casa de Vila'-> 2212
Nº de imóveis com área entre 60 e 100 metros quadrados, incluindo os limites -> 8719
Nº de imóveis que tenham pelo menos 4 quartos e aluguel menor que R$ 2.000,00 -> 41


##### <span style="color:pink; font-weight:bold"> Relatório de análise V</span>

Agora, identificaremos problemas no banco de dados e buscaremos soluções par eles. Especificamente, lidaremos com os missings, ou seja, vamos tratar os dados faltantes.

In [108]:
dados = pd.read_csv('dados/aluguel_residencial.csv', sep = ';')
dados

Tipo                Bairro  Quartos  Vagas  Suites  Area  \
0        Quitinete            Copacabana        1      0       0    40   
1             Casa       Jardim Botânico        2      0       1   100   
2      Apartamento                Centro        1      0       0    15   
3      Apartamento          Higienópolis        1      0       0    48   
4      Apartamento          Vista Alegre        3      1       0    70   
...            ...                   ...      ...    ...     ...   ...   
22575  Apartamento                 Méier        2      0       0    70   
22576    Quitinete                Centro        0      0       0    27   
22577  Apartamento           Jacarepaguá        3      1       2    78   
22578  Apartamento  São Francisco Xavier        2      1       0    48   
22579  Apartamento                Leblon        2      0       0    70   

        Valor  Condominio  IPTU  
0      1700.0       500.0  60.0  
1      7000.0         NaN   NaN  
2       800.0       390.0  20.0  
3       800.0       230.0   NaN  
4      1200.0         NaN   NaN  
...       ...         ...   ...  
22575   900.0       490.0  48.0  
22576   800.0       350.0  25.0  
22577  1800.0       800.0  40.0  
22578  1400.0       509.0  37.0  
22579  3000.0       760.0   NaN  

[22580 rows x 9 columns]

É possível ver que as variáveis condominio e iptu apresentam dados faltantes. Mas precisamos de um método pra auxiliar a analise do conjunto de dados inteiro.

Possuímos dois métodos que nos ajudam a realizar a seleção que precisamos. O primeiro deles é isnull(). Tal método irá gerar um DataFrame booleano, em que a observação marcada como True caracteriza um dado nulo, isto é, um missing. 

In [109]:
dados.isnull()

Tipo  Bairro  Quartos  Vagas  Suites   Area  Valor  Condominio   IPTU
0      False   False    False  False   False  False  False       False  False
1      False   False    False  False   False  False  False        True   True
2      False   False    False  False   False  False  False       False  False
3      False   False    False  False   False  False  False       False   True
4      False   False    False  False   False  False  False        True   True
...      ...     ...      ...    ...     ...    ...    ...         ...    ...
22575  False   False    False  False   False  False  False       False  False
22576  False   False    False  False   False  False  False       False  False
22577  False   False    False  False   False  False  False       False  False
22578  False   False    False  False   False  False  False       False  False
22579  False   False    False  False   False  False  False       False   True

[22580 rows x 9 columns]

o método notnull(), que funciona exatamente da maneira inversa ao isnull(): se a informação for nula, será utilizada a notação False. Contudo, ainda que marquemos true ou false nos dados, ainda não temos a capacidade de analisá-los integralmente.

In [110]:
dados.info() # exibe as informações do dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22580 entries, 0 to 22579
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Tipo        22580 non-null  object 
 1   Bairro      22580 non-null  object 
 2   Quartos     22580 non-null  int64  
 3   Vagas       22580 non-null  int64  
 4   Suites      22580 non-null  int64  
 5   Area        22580 non-null  int64  
 6   Valor       22571 non-null  float64
 7   Condominio  20765 non-null  float64
 8   IPTU        15795 non-null  float64
dtypes: float64(3), int64(4), object(2)
memory usage: 1.6+ MB


A segunda linha nos mostra a quantidade de registros disponíveis no DataFrame, no caso 22580. Para cada variável, também há um número de registros não nulos, isso nos ajuda a identificar os problemas e onde devemos aplicar soluções.

Iremos observar as variáveis que apresentam dados faltantes.

In [111]:
dados['Valor'].isnull()

0        False
1        False
2        False
3        False
4        False
         ...  
22575    False
22576    False
22577    False
22578    False
22579    False
Name: Valor, Length: 22580, dtype: bool

In [112]:
dados[dados['Valor'].isnull()]

Tipo           Bairro  Quartos  Vagas  Suites  Area  Valor  \
58    Apartamento  Barra da Tijuca        2      1       1    70    NaN   
1492  Apartamento             Leme        2      0       0    75    NaN   
1683         Casa     Campo Grande        3      4       3   363    NaN   
2012  Apartamento         Botafogo        2      0       0    95    NaN   
2034  Apartamento       Copacabana        2      0       0    72    NaN   
4941         Casa     Campo Grande        3      2       1   100    NaN   
8568  Apartamento             Leme        2      0       1    75    NaN   
8947  Apartamento           Glória        3      0       1   135    NaN   
9149  Apartamento            Gávea        3      1       1   105    NaN   

      Condominio   IPTU  
58         970.0   68.0  
1492       878.0    NaN  
1683         NaN    NaN  
2012      1010.0  170.0  
2034       850.0    NaN  
4941         NaN    NaN  
8568       878.0    NaN  
8947       910.0  228.0  
9149       880.0  221.0

A principal informação em nosso banco de dados são os valores dos aluguéis, portanto não faz sentido que tenhamos dados nulos neste campo. Precisamos elimina-los, e para tanto, usaremos o método dropna().

Criaremos uma variável A, que abrigara dados, isto é, nosso DataFrame original. Em seguida, escreveremos dados.dropna(), o método receberá como argumento susbset, que se trata de uma lista de variáveis. Em seguida, criaremos uma nova variável B, que abrigará o resultado do novo conjunto de dados que não contém valores nulos. Então, escreveremos A - B.

In [113]:
A = dados.shape[0]
dados.dropna(subset = ['Valor'])
B = dados.shape[0]
A - B

0

In [114]:
A = dados.shape[0]
dados.dropna(subset = ['Valor'], inplace = True)
B = dados.shape[0]
A - B

9

In [115]:
dados[dados['Valor'].isnull()]

Empty DataFrame
Columns: [Tipo, Bairro, Quartos, Vagas, Suites, Area, Valor, Condominio, IPTU]
Index: []

Temos as variáveis Condomínio e IPTU com alguns problemas. No caso de Condominio, os dados nulos nem sempre são incorretos, afinal no caso de uma casa, realmente não há esse tipo de tarifa. Para este caso, devemos considerar a variável Tipo: caso o imóvel seja um apartamento, os dados nulos serão excluídos.

In [116]:
dados[dados['Condominio'].isnull()].shape[0]

1813

In [117]:
selecao = (dados['Tipo'] == 'Apartamento') & (dados['Condominio'].isnull())

se nos atentarmos a nossa seleção, verificaremos que ela nos fornece exatamente as assinaturas que deverão ser removidas do banco de dados. Reaproveitaremos a estrutura já utilizada anteriormente neste notebook, e passaremos a seleção que criamos (selecao). Para eliminar as assinaturas nulas selecionadas ao invés de coletá-las, iremos adicionar ~ em selecao, que inverte a Series booleana.

In [118]:
A = dados.shape[0]
dados = dados[~selecao]
B = dados.shape[0]
A - B

745

In [119]:
dados[dados['Condominio'].isnull()].shape[0]

1068

O que faremos é manter esses dados, mas atribuir o valor 0 a eles. Temos uma função para esta ação fillna(), que receberá o valor 0 e inplance = True.

In [122]:
dados = dados.fillna({'Condomínio':0, 'IPTU':0})

In [123]:
dados[dados['Condominio'].isnull()].shape[0]

0

In [124]:
dados[dados['IPTU'].isnull()].shape[0]

0

In [125]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
Index: 21826 entries, 0 to 22579
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Tipo        21826 non-null  object 
 1   Bairro      21826 non-null  object 
 2   Quartos     21826 non-null  int64  
 3   Vagas       21826 non-null  int64  
 4   Suites      21826 non-null  int64  
 5   Area        21826 non-null  int64  
 6   Valor       21826 non-null  float64
 7   Condominio  21826 non-null  float64
 8   IPTU        21826 non-null  float64
dtypes: float64(3), int64(4), object(2)
memory usage: 1.7+ MB


In [127]:
dados.to_csv('dados/aluguel_residencial.csv', sep = ';', index = False)

In [128]:
dados

Tipo                Bairro  Quartos  Vagas  Suites  Area  \
0        Quitinete            Copacabana        1      0       0    40   
1             Casa       Jardim Botânico        2      0       1   100   
2      Apartamento                Centro        1      0       0    15   
3      Apartamento          Higienópolis        1      0       0    48   
5      Apartamento              Cachambi        2      0       0    50   
...            ...                   ...      ...    ...     ...   ...   
22575  Apartamento                 Méier        2      0       0    70   
22576    Quitinete                Centro        0      0       0    27   
22577  Apartamento           Jacarepaguá        3      1       2    78   
22578  Apartamento  São Francisco Xavier        2      1       0    48   
22579  Apartamento                Leblon        2      0       0    70   

        Valor  Condominio  IPTU  
0      1700.0       500.0  60.0  
1      7000.0         0.0   0.0  
2       800.0       390.0  20.0  
3       800.0       230.0   0.0  
5      1300.0       301.0  17.0  
...       ...         ...   ...  
22575   900.0       490.0  48.0  
22576   800.0       350.0  25.0  
22577  1800.0       800.0  40.0  
22578  1400.0       509.0  37.0  
22579  3000.0       760.0   0.0  

[21826 rows x 9 columns]